lib/whitehall_importer.rb

```diff
@@ -1,27 +1,31 @@
 # frozen_string_literal: true
 
 module WhitehallImporter
-  def self.import(whitehall_document)
-    record = WhitehallImport.create!(
+  def self.import_and_sync(whitehall_document)
+    whitehall_import = WhitehallImport.create!(
       whitehall_document_id: whitehall_document["id"],
       payload: whitehall_document,
       content_id: whitehall_document["content_id"],
       state: "importing",
     )
 
-    begin
-      Import.call(whitehall_document)
-      record.update!(state: "completed")
-    rescue StandardError => e
-      record.update!(error_log: e.message,
-                     state: "import_failed")
-    end
+    import(whitehall_import)
 
-    record
+    sync(whitehall_import) if whitehall_import.imported?
+
+    whitehall_import
+  end
+
+  def self.import(whitehall_import)
+    document = Import.call(whitehall_import.payload)
+    whitehall_import.update!(document: document, state: "imported")
+  rescue StandardError => e
+    whitehall_import.update!(error_log: e.message, state: "import_failed")
   end
 
-  def self.sync(document)
-    ResyncService.call(document)
-    ClearLinksetLinks.call(document.content_id)
+  def self.sync(whitehall_import)
```

Comment:
Should this and `.import` perhaps raise if the import model isn't in the expected state?

Reply:
Hmm I don't think these raises are needed due to import being called just after the WhitehallImport record is persisted with a state `importing`, if this for whatever reason failed then the code wouldn't get run anyway! 

As for the sync function, there is a conditional checking whether the thing has been imported before calling the function so again, the sync code would never run if it hasn't been imported. I can see the value if these functions were to be called separately but right now it seems overkill, what do you think?

Reply:
I think this depends on whether we consider these methods to be part of the public interface of WhitehallImporter or not.

We can mark them as private with `private_class_method` if they're just private. Though I'd expect we'd at least want sync available in case we want to try retry if a sync fails (Bruce and I have been talking about failures and retries today).

I anticipated we'd want the raises as a sense check since if a whitehall_import is passed to either of these in an unexpected state it causes data inconsistency. For example if you call .import with something that is already imported you'll get it set "import_failed" when that is not the case. Similarly if you pass a whitehall_import that is "importing" into .sync then it'll end up with a sync_failed status despite having never been imported.

Reply:
True, however we my want to think of the states that should execute the raise if we're thinking about this code being used beyond the scope of this piece of work. E.g for sync, I assume a state of `sync_failed` or `imported` should be allowed to pass through, but any of the rest should be raised as `sync_failed` could've occurred because of a transient publishing-api error.

Reply:
Have added the raises on the back of your suggestion!